### PR TITLE
feat: add increment & timeFilter for bitbucket

### PR DESCRIPTION
**Pull request by likyh**
### Summary
1. add transformationRule for bitbucket
2. add scope
3. support create bp with bitbucket in config-ui
4. support local bitbucket
5. support time filter
6. support increment
7. https://github.com/apache/incubator-devlake/issues/4294

### Does this close any open issues?
Related #4340 

### Screenshots

Config-ui:
image
image
image
image
image

CURD for transformationRule:
image
image
image

I have tested and confirmed that all columns in the tool layer tables have the proper value. Such as:
image

### CODE REVIEW

File: backend/plugins/bitbucket/api/remote.go [hunk 1]
```diff
@@ -0,0 +1,368 @@
+/*
```
Comment (likyh) [inline backend/plugins/bitbucket/api/remote.go:1]: this file imitates from https://github.com/apache/incubator-devlake/blob/3285a5aa58d8e678cd4ac1fedd06138d28fc23cc/backend/plugins/gitlab/api/remote.go

File: backend/plugins/bitbucket/api/blueprint_v200.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
+/*
```
Comment (likyh) [inline backend/plugins/bitbucket/api/blueprint_v200.go:1]: this file imitates from https://github.com/apache/incubator-devlake/blob/main/backend/plugins/github/api/blueprint_v200.go

File: backend/plugins/bitbucket/api/scope.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
+/*
```
Comment (likyh) [inline backend/plugins/bitbucket/api/scope.go:1]: this file imitates from https://github.com/apache/incubator-devlake/blob/main/backend/plugins/github/api/scope.go

File: backend/plugins/bitbucket/api/transformation_rule.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (likyh) [inline backend/plugins/bitbucket/api/transformation_rule.go:1]: this file imitates from https://github.com/apache/incubator-devlake/blob/main/backend/plugins/github/api/transformation_rule.go

File: backend/plugins/bitbucket/bitbucket.go [hunk 5]
```diff
@@ -30,17 +30,17 @@ var PluginEntry impl.Bitbucket //nolint
 func main() {
 	cmd := &cobra.Command{Use: "bitbucket"}
 	connectionId := cmd.Flags().Uint64P("connectionId", "c", 0, "bitbucket connection id")
-	owner := cmd.Flags().StringP("owner", "o", "", "bitbucket owner")
-	repo := cmd.Flags().StringP("repo", "r", "", "bitbucket repo")
+	fullName := cmd.Flags().StringP("fullName", "n", "", "bitbucket id: owner/repo")
```
Comment (likyh) [inline backend/plugins/bitbucket/bitbucket.go:33]: use one column for options.

File: backend/plugins/bitbucket/impl/impl.go [hunk 6]
```diff
@@ -79,6 +90,9 @@ func (p Bitbucket) SubTaskMetas() []plugin.SubTaskMeta {
 		tasks.CollectApiPrCommitsMeta,
 		tasks.ExtractApiPrCommitsMeta,
 
+		tasks.CollectApiCommitsMeta,
+		tasks.ExtractApiCommitsMeta,
```
Comment (likyh) [inline backend/plugins/bitbucket/impl/impl.go:94]: add unused tasks to allow open it on advance mod

File: backend/plugins/bitbucket/impl/impl.go [hunk 7]
```diff
@@ -166,3 +222,54 @@ func (p Bitbucket) Close(taskCtx plugin.TaskContext) errors.Error {
...
+	err = taskCtx.GetDal().First(&repo, dal.Where(
+		"connection_id = ? AND bitbucket_id = ?",
+		op.ConnectionId, op.FullName))
+	if err == nil {
+		if op.TransformationRuleId == 0 {
+			op.TransformationRuleId = repo.TransformationRuleId
+		}
+	} else {
+		if taskCtx.GetDal().IsErrorNotFound(err) && op.FullName != "" {
+			var repo *tasks.BitbucketApiRepo
```
Comment (likyh) [inline backend/plugins/bitbucket/impl/impl.go:246]: request repo from bitbucket api

**Review by warren830 — APPROVED** [review]
TL,DR

File: backend/plugins/bitbucket/api/remote.go [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+		// append group to output
+		for _, group := range resBody.Values {
+			child := RemoteScopesChild{
+				Type: TypeGroup,
+				Id:   group.Workspace.Slug,
+				Name: group.Workspace.Name,
+				// don't need to save group into data
+				Data: nil,
+			}
+			if gid != "" {
```
Comment (mappjzc) [inline backend/plugins/bitbucket/api/remote.go:170]: at the first level if,you have checked the gid already.

File: backend/plugins/bitbucket/api/remote.go [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+		// append repo to output
+		for _, repo := range resBody.Values {
+			child := RemoteScopesChild{
+				Type: TypeScope,
+				Id:   repo.FullName,
+				Name: repo.Name,
+				Data: tasks.ConvertApiRepoToScope(&repo, connection.ID),
+			}
+			child.ParentId = &gid
+			if *child.ParentId == "" {
```
Comment (mappjzc) [inline backend/plugins/bitbucket/api/remote.go:205]: At the first if else,you have checked the gid already. So the same as child.ParentId.

File: backend/plugins/bitbucket/api/remote.go [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+	}
+
+	// request search
+	query.Set("sort", "name")
+	query.Set("fields", "values.name,values.full_name,values.language,values.description,values.owner.username,values.created_on,values.updated_on,values.links.clone,values.links.self,pagelen,page,size")
+
+	gid := ``
+	if strings.Contains(s, `/`) {
+		gid = strings.Split(s, `/`)[0]
+		s = strings.Split(s, `/`)[0]
```
Comment (mappjzc) [inline backend/plugins/bitbucket/api/remote.go:299]: s = gid ? and set q = name~gid, and the url of get with gid? It's hard to understand.

**Review by mappjzc — APPROVED** [review]
remote-api part approve

File: backend/plugins/bitbucket/api/remote.go (hunk 10 above)
Comment (likyh) [reply backend/plugins/bitbucket/api/remote.go:299]: Bitbucket must search in a workspace(group). So here search text may like `aaa/searchText`

File: backend/plugins/bitbucket/api/remote.go [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+			}
+			if gid != "" {
+				child.ParentId = &gid
+			}
+
+			outputBody.Children = append(outputBody.Children, child)
+		}
+
+		// check groups count
+		if resBody.Size < pageData.PerPage {
```
[the remark this answers is not in the document]
Comment (likyh) [reply backend/plugins/bitbucket/api/remote.go:178]: resBody are not the same type

File: backend/plugins/bitbucket/api/remote.go (hunk 8 above)
Comment (likyh) [reply backend/plugins/bitbucket/api/remote.go:170]: fixed

File: backend/plugins/bitbucket/api/remote.go (hunk 9 above)
Comment (likyh) [reply backend/plugins/bitbucket/api/remote.go:205]: fixed

File: backend/plugins/bitbucket/api/blueprint_v200.go [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+			return nil, err
+		}
+
+		// add gitex stage
+		if utils.StringsContains(bpScope.Entities, plugin.DOMAIN_TYPE_CODE) {
+			cloneUrl, err := errors.Convert01(url.Parse(repo.CloneUrl))
+			if err != nil {
+				return nil, err
+			}
+			token := strings.Split(connection.Password, ",")[0]
```
Comment (warren830) [inline backend/plugins/bitbucket/api/blueprint_v200.go:138]: Can we use multiple tokens in Bitbucket?

Comment (likyh) [reply backend/plugins/bitbucket/api/blueprint_v200.go:138]: fixed

**Review by mintsweet — APPROVED** [review]
LGTM.

**Review by warren830 — APPROVED** [review]
LGTM

**Pull request merged**